File: day11/plutonian.pl
#!/usr/bin/env perl
use v5.40;
use autodie;
use List::Util qw(sum);

my @stones;
open my $fh, '<', $ARGV[0];
while (<$fh>) {
    chomp;
    @stones = split ' ';
}

my %cnts;
for my $stone (@stones) {
    $cnts{$stone}++;
}

for my $i (1..$ARGV[1]) {
    my %new_cnts;
    while (my ($k, $v) = each %cnts) {
        if ($k == 0) {
            $new_cnts{1} += $v;
        } elsif ((my $len = length($k)) % 2 == 0) {
            $new_cnts{0 + substr $k, 0, $len/2} += $v;
            $new_cnts{0 + substr $k, $len/2} += $v;
        } else {
            $new_cnts{$k * 2024} += $v;
        }
    }
    %cnts = %new_cnts;
    printf "i = %d sum = %d\n", $i, sum values %cnts;
}

File: day11/plutonian.py
import sys

def change(stone):
    if stone == 0:
        return [1]
    else:
        digits = str(stone)
        if (cnt := len(digits)) % 2 == 0:
            return [int(digits[0:cnt//2]), int(digits[cnt//2:])]
        else:
            return [stone * 2024]

def main():
    with open(sys.argv[1]) as f:
        for line in f:
            stones = [int(tok) for tok in line.rstrip().split(' ')]

    for i in range(int(sys.argv[2])):
        new_stones = []
        for stone in stones:
            new_stones += change(stone)

        stones = new_stones
        print(f'{i=} {len(stones)=}')

    print('Part 1:', len(stones))
        
main()
